clients/da-vinci-client/src/main/java/com/linkedin/davinci/client/DaVinciRecordTransformer.java

```diff
@@ -29,77 +36,161 @@ public abstract class DaVinciRecordTransformer<K, V, O> {
+   *
+   * @param key the key of the record to be transformed
+   * @param value the value of the record to be transformed
+   * @return the transformed value
+   */
+  public abstract O transform(Lazy<K> key, Lazy<V> value);
+
+  /**
+   * Implement this method to manage custom state outside the Da Vinci Client.
+   *
+   * @param key the key of the record to be put
+   * @param value the value of the record to be put,
+   *              derived from the output of {@link #transform(Lazy key, Lazy value)}
    */
-  public abstract O put(Lazy<K> key, Lazy<V> value);
+  public abstract void processPut(Lazy<K> key, Lazy<O> value);
 
   /**
-   * By default, deletes will proceed. This can be overridden if some deleted records should be kept.
+   * Override this method to customize the behavior for record deletions.
+   * By default, records will be deleted. Return a non-null object to keep the record in storage.
    *
-   * @param key to be deleted
+   * @param key the key of the record to be deleted
    * @return the object to keep in storage, or null to proceed with the deletion
    */
-  public O delete(Lazy<K> key) {
+  public O processDelete(Lazy<K> key) {
     return null;
   }
 
   /**
-   * This function is called as a lifecycle event at the beginning of an ingestion task.
+   * Lifecycle event triggered before records are consumed.
+   * Use this method to perform setup operations such as opening database connections or creating tables.
    *
    * By default, it performs no operation.
-   *
    */
-  public void onStartIngestionTask() {
+  public void onStart() {
     return;
   }
 
   /**
-   * This function is called as a lifecycle event at the end of an ingestion task
+   * Lifecycle event triggered after records are done consuming.
+   * Use this method to perform cleanup operations such as closing database connections or dropping tables.
    *
    * By default, it performs no operation.
-   *
    */
-  public void onEndIngestionTask() {
+  public void onEnd() {
     return;
   }
 
   /**
-   * Takes a value, serializes it and wrap it in a ByteByffer.
+   * Transforms and processes the given record.
    *
-   * @param schema the Avro schema defining the serialization format
-   * @param value value the value to be serialized
-   * @return a ByteBuffer containing the serialized value wrapped according to Avro specifications
+   * @param key the key of the record to be put
+   * @param value the value of the record to be put
+   * @return the transformed record
    */
-  public final ByteBuffer getValueBytes(Schema schema, V value) {
-    ByteBuffer transformedBytes = ByteBuffer.wrap(new AvroSerializer(schema).serialize(value));
-    ByteBuffer newBuffer = ByteBuffer.allocate(Integer.BYTES + transformedBytes.remaining());
-    newBuffer.putInt(1);
-    newBuffer.put(transformedBytes);
-    newBuffer.flip();
-    return newBuffer;
+  public final O transformAndProcessPut(Lazy<K> key, Lazy<V> value) {
```

Comment:
Reading about the behavior of the key and value in [SIT](https://github.com/kvargha/venice/blob/91ba18590c7b742b6b73ced7e351602c705740b9/clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java#L3316-L3321), I don't think value needs to be lazy since the operation of assembling has already occurred. Right now, it doesn't really do anything.

Thoughts @ZacAttack?

Reply:
I'm not sure that's exactly right is it?  The point of the lazy is to handle not having to do value deserialization or chunking unless the user needs it.  The code today acts like a follower node should which is just storing key/values to bytes, and the point of the lazy is to give users the ability to read the value as a serialized/assembled record if they need to process it for their logic.

Deserializing twice is a problem.... but could be solved with a lazy.

Reply:
> Deserializing twice is a problem.... but should be solved via the lazy right? The lazy function should deserialize once and then on successive calls to get it will just return a cached value.
Yes, you're right.

Reply:
Added a todo to make chunking lazy.

Reply:
Just to document the discussion made during the meeting. We need to think about what to do with the lazy value, since the parameter for transform is lazy, and processPut is also lazy. Should transform return a lazy value, or should we create a lazy/non-lazy API? TBD.

Reply:
To document this a bit further, the 4 scenarios we need to make sure we have correctness for our:

1. User deserialized the value and altered it
2. User deserialized the value and didn't alter it
3. User did not deserialize the value and just passed it long to the storage engine
4. User did not deserialize the value and passed along something else entirely